docs/typescript.md

```diff
@@ -71,17 +71,31 @@ const App = () => (
 ```tsx
 import styled from 'react-emotion';
 
-const NotALink = styled('div')`
+const NotALink = styled<{}, 'div'>('div')`
   color: red;
 `
 
 const Link = NotALink.withComponent('a')
 
+type TextProps = {
+  className: string;
+  text: string;
+}
+const Text = ({ className, text }: TextProps) => <div className={className}>{text}</div>
+const StyledText = styled<{}, TextProps>(Text)`
+  fontSize: 20px;
+  color: white;
+`;
+
+const OtherLink = StyledText.withComponent('a');
+
 const App = () => <Link href="#">Click me</Link>
 
 // No errors!
 ```
 
+When you use `withComponent`, you should explicitly provide the type of props used to make style. If you don't, TS will think all props for `Text` are used to make style, and request user to give all those props when using the result component of `withComponent`.
```

Comment:
Could you elaborate on `TS will think all props for Text are used to make style`, I don't really understand what it means.

Reply:
@mitchellhamilton I mean `TS will consider all props for Text as props to make style`. Is this sentence more sensible?

Reply:
No, I still don’t understand it. I don’t understand what you mean by `as props to make style`. how do the props `make` the style?

Reply:
@mitchellhamilton I mean something like following
```
(props) => ({
  backgroundColor: props.bgColor
})
```
(`bgColor` props is used to make style)

Reply:
For example,
1. When you have `Text` component that requires `bgColor`, `content`, you can make following styled component.
    ```
    const StyledText = styled(Text)({
      fontSize: '50px',
    }, props => ({
      backgroundColor: props.bgColor,
    }));
    ```
    In this `StyledText`, `bgColor` props is used to make style, but `content` isn't.
2. Now, suppose that you want to make a new styled component with different inner (base?) component `Link`, by using `withComponent`. Suppose that `Link` requires only `target` as a props. You probably do something like following.
    ```
    const StyledLink = StyledText.withComponent(Link);
    ```
3. However, the problem occurs in TypeScript. TypeScript cannot determine which props are used to make style, i.e., it cannot find `bgColor` is the only props used in style of `StyledText`. Therefore it should do one of following things.
    1. Suppose all props are used in style
    2. Suppose no props are used in style

    However, if you choose first one, TypeScript request to use `bgColor`, `content`, `target` to `StyledLink`, even if `Link` does not require `content`. This is why I choose second one, but legacy typing choose first one and some user complain about 'Why are there no props inference?'.

Reply:
@mitchellhamilton Glamorous do something with second choice (not exactly same, but similar), and `styled-components` have something similar with first one, but with some wrong typed `withComponent`.

Reply:
@tkh44 I just changed docs. Could you check this again?